### ISSUE TITLE
Can import Google calendar, but not sync

### ISSUE THREAD

**Issue by janaxhell** [MarcelRobitaille/nextcloud_google_synchronization#8]
I'm on v25.0.6 inside AIO. My domain passes through Nginx Proxy Manager.
I have setup the Google Synchronization following this [video](https://www.youtube.com/watch?v=HQOQ2NCUuVQ&t=609s)
then I moved to Personal Settings and connected to my Google account.

I can import everything, but cannot sync. After importing, Nextcloud calendar populated with all Google events, but when I tried to add a couple of new test events to Google, they did not sync to Nextcloud. When I instead clicked Import again, they appeared.

The logging is full of these errors repeated over and over https://pastebin.com/9SRWmdSz

As a side note, I cannot sync from DAVx5 app either, despite permission granted from Nextcloud android app.
So my Nextcloud can see and import the contents of my Google account, but not sync. What's wrong?

Doing more tests, I can now sync between

**Nextcloud** <---> **DAVx5** <---> **Simple Calenader**

but not Google Calendar in any direction, only import

**Comment by lroyb** [comment]
if you read the markdown file, it says that it is a fork of the 'google intergration' app. Most importantly it says that ' it should not be used as well as'. So, I disabled the google intergration app, leaving the google Synchronization app enabled. and it worked first try afterward.

**Comment by janaxhell** [comment]
I posted this 4 months ago, I have updated NC several times since and i'm on Nextcloud Hub 5 (27.0.2). 
As of now I cannot find Google Integration anywhere, but DAVx5 and Google Synchronization still give errors all the time like this

`Failed to get number of Google photos: "ServerException|ClientException, message:Client error: `GET 
 https:\/\/photoslibrary.googleapis.com\/v1\/albums?pageSize=50` resulted in a `401 Unauthorized` response:\n{\n \"error\": 
 {\n \"code\": 401,\n \"message\": \"Request had invalid authentication credentials. Expected OAuth 2 access (truncated...)\n 
 status code: 401"`

I got tired and just ignore it.

**Comment by lroyb** [comment]
By the look from the above message, have you created the oauth2 app with the correct Google api's enabled. Then successfully logged into your Google account through nextcloud, which you can then give nextcloud access to the oauth2 app?

**Comment by janaxhell** [comment]
Of course, but it periodically forgets about it and wants to grant access again. That is so annoying I gave up.

**Comment by MarcelRobitaille** [comment]
Hi everyone. Sorry I haven't been around. For some reason, even though I own this fork, it was not set up to send me notifications on issues and PRs. Sorry about that! I've signed up now.

Sorry you are experiencing this issue. I will look into it.

**Comment by MarcelRobitaille** [comment]
Could you please try [the latest release](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/releases/tag/v2.1.0-1) and let me know if you are still experiencing this.

**Comment by janaxhell** [comment]
@MarcelRobitaille 

Hi, I have just installed and tried the latest version available. This is what happened:
- Sync replied "Successfully registered background job", but nothing really happened
- Import imported already existing events, but as duplicates 1 hour earlier than correct one. But from what I seem to understand, not all of them, only those that I added as repeating events before Solar Time change in autumn. Those that I added after have the correct time.

No error message from Google, but now I have Nextcloud Calendar populated with tons of wrong duplicate events.
The events with double entries cyan/blue in the picture have the wrong time in the cyan block and right time in the blue block.

![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/55061479/b8157788-cf0c-4ebd-bde1-18422a0ce711)

**Comment by MarcelRobitaille** [comment]
@janaxhell what happens when you press the import button? Are there any logs you could share?

**Comment by janaxhell** [comment]
I use Nextcloud AIO, from which service do you want me to check the logs?

nextcloud-aio-apache
nextcloud-aio-collabora
nextcloud-aio-database
nextcloud-aio-fulltextsearch
nextcloud-aio-imaginary
nextcloud-aio-mastercontainer
nextcloud-aio-nextcloud
nextcloud-aio-notify-push
nextcloud-aio-redis

The nextcloud-aio-nextcloud service is showing right now this:

```
2024-02-29 17:29:55 Importing it.italian#holiday@group.v.calendar.google.com... done. Added 
2024-02-29 17:29:56 Importing XXXXXX@XXXXXX... done. Added 0
2024-02-29 17:30:08 Importing XXXXXX@XXXXXX... done. Added 
2024-02-29 17:34:55 Importing XXXXXX@XXXXXX... done. Added 0
2024-02-29 17:34:59 Importing XXXXXX@XXXXXX... done. Added 
2024-02-29 17:34:59 Importing it.italian#holiday@group.v.calendar.google.com... done. Added 
2024-02-29 17:39:55 Importing XXXXXX@XXXXXX... done. Added 0
2024-02-29 17:39:59 Importing it.italian#holiday@group.v.calendar.google.com... done. Added 
2024-02-29 17:39:59 Importing XXXXXX@XXXXXX... done. Added 
2024-02-29 17:44:55 Importing XXXXXX@XXXXXX... done. Added 0
2024-02-29 17:44:58 Importing it.italian#holiday@group.v.calendar.google.com... done. Added 
2024-02-29 17:44:59 Importing XXXXXX@XXXXXX... done. Added 
2024-02-29 17:50:04 Importing it.italian#holiday@group.v.calendar.google.com... done. Added 
2024-02-29 17:50:04 Importing XXXXXX@XXXXXX... done. Added 0
2024-02-29 17:50:08 Importing XXXXXX@XXXXXX... done. Added 
2024-02-29 17:54:55 Importing it.italian#holiday@group.v.calendar.google.com... done. Added 
2024-02-29 17:54:55 Importing XXXXXX@XXXXXX... done. Added 0
```
but in the NC GUI I haven't enabled the Sync thick for hliday/festivity, only my own gmail address.

**Comment by MarcelRobitaille** [comment]
Thank you for the logs. Just to be very clear because it's hard to tell from your language:

Please press this button:
![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/8503756/7e143aec-a803-4dca-be3d-6f3b6e00c7ce)

You should see something like this in the top right:
![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/8503756/d1becb8d-2a6b-4fd3-bdd4-a53c2f470434)

For me, everything is up to date, so I see '0 events successfully imported".
You can think of the button "Sync calendar" as automatically pressing "Import calendar" every 5 minutes. "Import calendar" is easier to troubleshoot, so that's why I want to see if it works first.

**Comment by janaxhell** [comment]
Yes, that appears to me as well and the first time I did it today I saw a number different from zero, like 24 events added:

![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/55061479/b4049e69-4f6d-49c9-b1ea-383659684471)

But it is incoherent with what I see:

![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/55061479/aabe54f3-c05b-4671-9615-c0f7c46d4bd8)

On the left side panel you see that Personal (= my gmail account) is blue and Baikal (external caldav) is cyan. When I Import from gmail, I see cyan blocks being added instead of the blue ones. I have configured Google Sync to use my gmail account, not Baikal, that's to sync the calendar on my phone. Maybe I should completely clear Nextcloud calendar and try gain from scratch, but I'm going offline now and won't be able to try again before tomorrow afternoon.

**Comment by MarcelRobitaille** [comment]
Thanks for the updates. That's weird. It should create a dedicated calendar for each calendar you sync from google. You should have a new calendar blah@google.com. It should not modify Personal or Baikal at all. If it does, that's very bad.

Could you please press the "Delete all background jobs" button in the admin page?

**Comment by MarcelRobitaille** [comment]
So everything works correctly except for dailight savings time? Is it possible your nextcloud locale / time zone settings are misconfigured?

**Comment by MarcelRobitaille** [comment]
Thanks for the update. Is it a recurring event in Google Calendar or is each event entered individually? Could you please let me know which time zone you have selected in Google and in Nextcloud?

If it's not too much trouble, could you please uninstall this app and install https://github.com/nextcloud/integration_google to see if the issue also occurs there?

**Comment by janaxhell** [comment]
I've set a few events in the first week, then set them to repeat every week until X. 
In Google Calendar I've set GMT+1 and I have supposedly set the same in NC, but I cannot find that setting anywhere,
I have disable Google Synchronization and Installed Google Integration (that's all I can find in the Apps menu), but the version number does not match the one in the github you linked: github = 1..0.9, Nextcloud = 2.1.0. Nevertheless if I click Visit Website in the appa details tab, I'm brought to the same github page thats shows v 1.0.9. This is quite confusing.
Also now I cannot see any entry in da admin column to connect to Google.

**Comment by MarcelRobitaille** [comment]
That is weird. This app is a fork of that one. The settings should mostly be the same.

About the versions, probably they just did not release the latest version yet. Don't worry about it, not much has changed.

**Comment by MarcelRobitaille** [comment]
The time zone should be here (don't ask me why)
![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/8503756/90419aec-07e5-4d37-80d3-b02ddc124ede)

The reason I'm asking so much about time zones is that all of the backend should be in UTC. It does not care about time zones or daylight savings or anything like that. It's up to the calendar app to read the UTC time from the database and display it properly.

> GMT+1

This is *very* interesting. GMT does not observe daylight savings as far as I know. The UK uses GMT only in the winter. They use BST (British summer time) in the summer. Could you please try setting your time zone in Google and Nextcloud to "Europe - Rome"?

If Nextcloud is set up to use daylight savings (like "Europe - Rome") and Google is not (GMT+1), this could be the cause of the shift.

**Comment by janaxhell** [comment]
Thanks, NC is set as Europe - Rome, Google Calendar is set as Primary time zone = (GMT+01:00) Central European Time - Rome.
I cannot see any option to set in a different way than GMT+x, there is no "Europe - Rome" entry .

![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/55061479/ec10de7b-6296-4e65-986f-907a9f674c41)

**Comment by MarcelRobitaille** [comment]
Ok, I am able to reproduce this with those timezone settings. I will take a look.

**Comment by MarcelRobitaille** [comment]
@janaxhell Are you the owner of the Google Calendar? You might be better served by importing it a different way (please see the disclaimers I added to the readme)

**Comment by janaxhell** [comment]
I won't be able to check this before the weekend, I'm commenting just to let you know I've not disappeared.

**Comment by janaxhell** [comment]
Here I am. Of course I am the owner of the calendar, I don't understand the rest of your post: why would I want to synchronize from NC to Google? Or is the ical method a workaround for the wrong event-times?

My scenario is: I want to move out from Google, but until I haven't found a reliable combination of apps and services to replace it, I prefer keeping both Google and NC. So I'd like to have a way to sync anything that I add to Google -> to Nextcloud. And then at some point get rid of Google. One unrelated thing that I'm missing for instance is a NC UI that has both events and tasks: in NC those are separated and afaik cannot be directly synced.

**Comment by MarcelRobitaille** [comment]
Hi. If you are the owner, there may be an easier way. What I described earlier is a different method to synchronize Google to Nextcloud, not a method to sync Nextcloud to Google. I have tested it on my side. I am able to reproduce your bug with this app, but the other method I described shows the time properly.

Here are the steps for the ical method. Sorry for not explaining it well earlier:
1. In the Google Calendar, click the 3 dots next to the calendar you want to import and go to settings.
2. Scroll down until the "Secret address in iCal format" and copy it
3. In Nextcloud Calendar, click "New calendar" and click "New subscription from link (read only)". Paste the secret link from before

I originally made this fork because the calendar I wanted to sync was owned by a team. I did not have access to get the private address. That's the advantage of this app over the ical method.

I'm not very familiar with Google Tasks, but Nextcloud Tasks seems good enough for me. If you assign a due date to a task, it will show up in your calendar. What are you trying to import? Tasks?

Regarding the edit:
> In the Google data migration tab there is no Sync Calendar button, only Import
Is it possible you have an app called "Google Integration" installed (from which this app is forked)?
> which still imports with the wrong times
Interesting. For me it's working. Is it possible you're using the "Google Integration" app instead of the ical method?

**Comment by janaxhell** [comment]
> Here are the steps for the ical method.

I already did that and it imported correctly. I cannot sync.

>> In the Google data migration tab there is no Sync Calendar button, only Import

> Is it possible you have an app called "Google Integration" installed (from which this app is forked)?
> which still imports with the wrong times
> Interesting. For me it's working. Is it possible you're using the "Google Integration" app instead of the ical method?

I had Google Synchronization at the beginning of this conversation, the you told me:
> If it's not too much trouble, could you please uninstall this app and install https://github.com/nextcloud/integration_google 

So since then I have always used Google Integration.. Which does not appear in the Admin menu. Should I replace it back? 

If it's not possible to synchronize with correct time, I suppose I'll just stay with Import and when it's possible to have a full monolithic Google Calendar+Tasks replacement, I'll move to that. 

> I'm not very familiar with Google Tasks, but Nextcloud Tasks seems good enough for me. 

It's not at all for me: it doesn't show tasks in calendar despite having it enabled to do so and it doesn't allow to set periodical/repeated events, I am trying jtxBoard for that, but it's only from phone, not convenient being mostly at my desktop.
Google Tasks is so easy: I just click any point on the Calendar, choose Task and set it as I wish. Nextcloud has a byzantine interface for Tasks.

**Comment by MarcelRobitaille** [comment]
> I already did that and it imported correctly. I cannot sync.

As far as I know, that method should also keep it up to date. What do you mean you cannot sync.

> I had Google Synchronization at the beginning of this conversation, then you told me:
> 
> >    If it's not too much trouble, could you please uninstall this app and install https://github.com/nextcloud/integration_google

Oh yea, I did tell you to do that, didn't I. Sorry about that.

Yes, repeating tasks are a big pain in NC. I'm also subscribed to that GH issue.

**Comment by janaxhell** [comment]
Ahh, so you mean that using the ical link will work as IF I click on the Sync button?
So the best thing now is to reinstall Google Synchronization AND use the ical link method, right?

**Comment by MarcelRobitaille** [comment]
I think you don't need any Google Synchronization / Google Integration app at all. I think you only need the ical sync method

**Comment by janaxhell** [comment]
Ok, thank you for all the help!

**Comment by janaxhell** [comment]
Sorry, one last thing: I've just tried to add a TEST event to Google Calendar. How/when should it appear in NC Calendar? I don't see anything else than the secret ical link to set.. Is there a timeframe after which ical link is synced?

**Comment by MarcelRobitaille** [comment]
I'm not sure. I don't think it's configurable anywhere. At a guess, when your background jobs run.

Administration > Basic Settings
![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/8503756/a4181074-db47-415a-b862-3b3d3ea6d625)

**Comment by janaxhell** [comment]
Yes, you are right: in that section it says "Last job ran 3 minutes ago." and in fact while you were replying, the TEST event appeared in Nextcloud. 
Thanks again!

**Comment by MarcelRobitaille** [comment]
Sounds like it's resolved. Closing this.

**Comment by janaxhell** [comment]
No, it's not, I checked again several other times and it did not sync anymore. Cron job shows as fired X minutes ago, but calendar does not sync. I gave up. It's not a vital feature for me, too many problems. Thanks anyway.

**Comment by MarcelRobitaille** [comment]
@janaxhell Thanks for the feedback. A lot of problems were discussed in this thread. Can you please let me know what problem you are still experiencing?

**Comment by janaxhell** [comment]
It does not sync. It did last time we talked, it stopped doing so sometime in the middle. If I add an event to GC, it never appears in NC.

**Comment by MarcelRobitaille** [comment]
Can you please check in the admin settings that the background jobs are being run correctly?

**Comment by janaxhell** [comment]
"Secondi fa" means "a few seconds ago"

![image](https://github.com/MarcelRobitaille/nextcloud_google_synchronization/assets/55061479/bae71b04-c02c-4012-a68f-86932c13274c)

**Comment by MarcelRobitaille** [comment]
Very nice internationalization :laughing: 

Sorry for asking the same questions again, but a lot of time passed and maybe something changed:
- What version of the app are you using?
- Could you please try pressing the button "Delete all background jobs" in "Connected accounts" and register the job again
- If this doesn't work, could you please share your serve logs.

**Comment by janaxhell** [comment]
> Very nice internationalization 😆
> 
> Sorry for asking the same questions again, but a lot of time passed and maybe something changed:
> 
> * What version of the app are you using?
Google integration
Version 2.2.0

> * Could you please try pressing the button "Delete all background jobs" in "Connected accounts" and register the job again

As usual I had to disconnect from Google and reconnect. As I clicked "Import", it imported all the events. 
Anyway as far as I remember this had nothing to do with ICAL job, which should be a separate automation.
In the end neither work after some time.

**Comment by Andrey007-Alt** [comment]
Hi!
Ubuntu 24.04.1 LTS
Nextcloud Hub 9 (30.0.5)
DB type MariaDB
PHP 8.3.6. (CLI)

config.php:
'dbtype' => 'mysql',
'version' => '30.0.5.1',
'overwrite.cli.url' => 'http://192.168.0.103/',
'dbname' => 'nextcloud',
'dbhost' => 'localhost',
'dbport' => '',
'dbtableprefix' => 'oc_',
'mysql.utf8mb4' => true,
'dbuser' => 'nextcloud',

There are many events in the logs:

![Image](https://github.com/user-attachments/assets/36eb5f3d-1a8a-4cf0-97e7-b58b25c005b7)

[google_synchronization] Warning: Error when creating calendar event "<redacted>" An exception occurred while executing a query: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect string value: '\xD1' for column `nextcloud`.`oc_calendarobjects_props`.`value` at row 1
	from ? by -- at 30 янв. 2025 г., 14:45:10

There seems to be a solution, everything is done according to the instructions.:
https://docs.nextcloud.com/server/latest/admin_manual/configuration_database/mysql_4byte_support.html

But it didn't help, what should I do?

**Comment by MarcelRobitaille** [comment]
@Andrey007-Alt This seems to be a separate problem. Please open a new issue.

Please clarify if this occurs when you try to import the calendar or if you've set it up to sync in the background.

**Comment by Andrey007-Alt** [comment]
> [@Andrey007-Alt](https://github.com/Andrey007-Alt) This seems to be a separate problem. Please open a new issue.
> 
> Please clarify if this occurs when you try to import the calendar or if you've set it up to sync in the background.

The first time I started it manually, all the events were loaded and displayed correctly in the calendar. Then the background synchronization is performed and these errors are already going through them.

**Comment by kevinstadler** [comment]
Hello,

I have the same problem as the original poster(s), connecting the Google account and manually importing a Calendar works, however syncing does not. There are no error messages in the log, it looks like the background task is simply never executed? We use webcron in our setup and it usually works well, when I call cron.php manually it gives `{"status":"success"}` however no sync is executed. I have also already tried 'Delete all background jobs', but nothing changes.

I'm on NC 31.0.6 and the latest Google Synchronization. Any ideas? I'd be happy to dig into the codebase myself if I knew where to start, are there any quirks or changes with NC's background job API that might be the cause of this?

**Comment by MarcelRobitaille** [comment]
@spongebyte Please open a new issue. This is an old issue, and there was just a new release, so it's very likely not the same root cause.

What do you mean "I simply cannot select the "sync" button"? Do you mean the button is disabled? Does nothing? Shows an error?

Since you said you recreated everything and you think something persisted, maybe you have to click "Delete all background jobs"

<img width="1427" height="1211" alt="Image" src="https://github.com/user-attachments/assets/685e4f7e-9ae9-4a85-874b-bd5441aea090" />

**Comment by lustsazeus-lab** [comment]
Hi @janaxhell, I reviewed this issue and can help with a 24-48h fast-track fix + regression test coverage.

Context noted: Google Calendar sync broken in self-hosted Nextcloud; likely wants fix.

If helpful, share version + minimal repro and I can post a concrete patch plan next.

If you want expedited implementation PR turnaround, I can also take it as a paid fast-track.

**Comment by lustsazeus-lab** [comment]
Given import works but incremental sync stalls, this often points to sync-token state failing while full-import path still succeeds.\n\nCould you share one recent log block around token refresh + scheduler run (plus Nextcloud/app/PHP versions)? That should quickly separate Google API refresh handling from cron/scheduler issues.

**Comment by MarcelRobitaille** [comment]
@janaxhell are you still experiencing this?

**Comment by janaxhell** [comment]
I stopped using this years ago..

**Comment by MarcelRobitaille** [comment]
Then I'll close this
